core/src/main/java/org/axonframework/config/MessageMonitorFactoryBuilder.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2010-2017. Axon Framework
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.config;
+
+import org.axonframework.common.Assert;
+import org.axonframework.messaging.Message;
+import org.axonframework.monitoring.MessageMonitor;
+import org.axonframework.monitoring.NoOpMessageMonitor;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.function.BiFunction;
+
+/**
+ * Class used by {@link DefaultConfigurer} to maintain the configuration for Message Monitors and create the function
+ * used in the {@link Component}.
+ */
+class MessageMonitorFactoryBuilder {
+
+    // Comparator to ensure the keys are processed in a reasonably stable order in #getFactoryForType
+    // Added comparing on the class' classloader as a class can be loaded multiple times by different classloaders
+    private Comparator<Class<?>> classComparator =
+            Comparator.comparing((Class<?> c) -> c.getName())
+                      .thenComparingInt((Class<?> c) -> c.getClassLoader().hashCode());
+
+    private Map<String, SortedMap<Class<?>, MessageMonitorFactory>> forNameFactories = new HashMap<>();
+    private SortedMap<Class<?>, MessageMonitorFactory> forTypeFactories = new TreeMap<>(classComparator);
+    private MessageMonitorFactory defaultFactory = (configuration, type, name) -> NoOpMessageMonitor.instance();
+    private boolean built = false;
+
+    MessageMonitorFactoryBuilder add(Class<?> componentType, String componentName, MessageMonitorFactory messageMonitorFactory) {
```

Comment:
Curious: Why are all these functions package private?

Reply:
Discussed this with @abuijze as well. This class is intended to be used only by the implementation of the framework. Hence the class is package private, as well as all its functions.

Reply:
Sounds fair!